Fix addressing for strided explicit copy operations

fknorr opened the pull request:
Data in host buffers should be assumed to be contiguous instead of mirroring the shape of the device accessor.

This also improves unit_tests/explicit_buffer_copy_host_ptr to account for strided copies.

illuhad commented on include/hipSYCL/sycl/handler.hpp:
```diff
@@ -437,7 +437,7 @@ void set_args(Ts &&... args);
                                         data_src};
     // Assume dest element size and allocation shape is the same as src
```
I think that this comment is now not correct anymore. Can you adjust or remove it?

illuhad commented on include/hipSYCL/sycl/handler.hpp:
```diff
@@ -471,7 +471,7 @@ void set_args(Ts &&... args);
 
     // Assume same element size and allocation shape as for dest
```
same here

illuhad commented on tests/unit_tests.cpp:
```diff
@@ -985,53 +985,114 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(explicit_buffer_copy_host_ptr, _dimensions,
 
-  std::shared_ptr<size_t> host_buf_result(new size_t[buf_size.size()],
-    std::default_delete<size_t[]>());
+  // modify each buffer on the device to avoid potential transfer elision
```
Is this really necessary? As soon as a host pointer is involved, I don't think that a SYCL implementation could elide an explicit copy operation as we cannot know what has happened to the host pointer. hipSYCL will also never elide explicit copies between accessors ("you asked explicitly for a copy, so you are getting one").

fknorr replied:
Since up to the first synchronization point in line 1090 (`queue.wait`) the execution is asynchronous, the host memory cannot be modified from user code without causing a race with the async copy. I'd argue an implementation could potentially detect that this host -> device -> host round trip is unobservable and replace it with a memcpy. Not that I see any reason why hipSYCL should ever implement that kind of logic.

I don't feel strongly about this, if you think it's unnecessary I'll remove it. Your call.

illuhad replied:
I also don't have strong feelings about this, but I'm leaning more towards removing. After all, these are the hipSYCL unit tests, so we can rely on guarantees that hipSYCL makes.

It's an interesting point you make. I think in order to observe that the host->device->host copy is unecessary the implementation would need to follow a lazy submission approach and evaluate the DAG only at the synchronization point. My understanding is that in SYCL 1.2.1 a `submit()` should always correspond to an OpenCL command submission right away. (With the new runtime hipSYCL is moving in the direction of lazy submission).

illuhad commented:
Can you please resolve merge conflicts and remove the unnecessary buffer modification in the tests, so that this PR can be merged?

fknorr commented:
Removed the buffer modification and rebased, but the `sycl_test` does not pass. I believe this is unrelated to the changes in this PR though.

CPU:
```
Running 107 test cases...
../sycl/sycl_test_suite.hpp(50): fatal error: in "explicit_copy_tests/explicit_buffer_copy_two_accessors_h2h<mpl___integral_c<int_ 2>>": critical check a[0] == b[0] has failed
unknown location(0): fatal error: in "usm_tests/allocations_in_kernels": memory access violation at address: 0x00000000: no mapping at fault address
../sycl/usm.cpp(176): last checkpoint: "allocations_in_kernels" test entry

*** 2 failures are detected in the test module "hipsycl unit tests"
```

CUDA/sm_75:
```
...
../sycl/usm.cpp(380): error: in "usm_tests/memset": check host_mem[i] == 12 has failed [0 != 0xc]
../sycl/usm.cpp(380): error: in "usm_tests/memset": check host_mem[i] == 12 has failed [0 != 0xc]
../sycl/usm.cpp(380): error: in "usm_tests/memset": check host_mem[i] == 12 has failed [0 != 0xc]
../sycl/usm.cpp(380): error: in "usm_tests/memset": check host_mem[i] == 12 has failed [0 != 0xc]
../sycl/usm.cpp(380): error: in "usm_tests/memset": check host_mem[i] == 12 has failed [0 != 0xc]
============== hipSYCL error report ============== 
hipSYCL has caught the following undhandled asynchronous errors: 

   0. from runtime/cuda/cuda_queue.cpp:217 @ submit_memcpy(): cuda_queue: Couldn't submit memcpy (error code = CUDA:1)

The application will now be terminated.
unknown location(0): fatal error: in "usm_tests/memset": signal: SIGABRT (application abort requested)
../sycl/usm.cpp(378): last checkpoint
terminate called recursively

*** 28675 failures are detected in the test module "hipsycl unit tests"
```

illuhad commented:
Thanks! The CPU one I think I've seen before, but all those CUDA errors I cannot reproduce. Can you please post the full output somewhere? For CUDA it's always important to look at the first error as CUDA errors can be sticky and cause subsequent operations to not return `cudaSuccess` (we should probably clean the errors somewhere inside hipSYCL in the future..)

fknorr commented:
I've created an issue with the full log.

illuhad commented:
Thank you!
